graphql_compiler/compiler/emit_sql.py

```diff
@@ -239,19 +340,152 @@ def start_global_operations(self):
             sql_expression = sqlalchemy.or_(sql_expression,
                                             self._came_from[self._current_alias].is_(None))
         self._filters.append(sql_expression)
 
+    def fold(self, fold_scope_location):
+        """Begin execution of a Fold Block."""
+        # Subquery will look as follows:
+        #
+        # SELECT
+        #   OuterVertex.uuid <- this value is determined from the edge descriptor
+        #   array_agg(InnerVertex.fold_output_column) AS fold_output
+        # FROM OuterVertex
+        # INNER JOIN ... <- INNER JOINs compiled during Unfold block
+        # ON ...
+        #          ...
+        # INNER JOIN InnerVertex <- INNER JOINs compiled during Unfold block
+        # ON ...
+        # GROUP BY OuterVertex.uuid
+
+        if self._in_fold:
+            raise AssertionError(u'Reached an invalid fold state.')
+
+        # begin the fold
+        self._in_fold = True
+        # outer vertex is whatever vertex occurs immediately before the fold
+        self._fold_outer_vertex_location = self._current_location
+        self._fold_inner_vertex_location = fold_scope_location
+
+        # basic info about the outer vertex and inner vertex
+        inner_vertex_table = self._sql_schema_info.vertex_name_to_table[
+            self._ir.query_metadata_table.get_location_info(fold_scope_location).type.name
+        ].alias()
+        outer_vertex_table = self._current_alias.alias()
+        inner_vertex_type = self._ir.query_metadata_table \
+            .get_location_info(fold_scope_location) \
+            .type \
+            .name
+        outer_vertex_type = self._current_classname
+
+        self._aliases[
+            (self._current_location.query_path, self._fold_inner_vertex_location)
+        ] = inner_vertex_table
+
+        # collect edge information to join the fold subquery to the main selectable
+        edge_direction = fold_scope_location.fold_path[0][0]
+        edge_name = fold_scope_location.fold_path[0][1]
+        full_edge_name = '{}_{}'.format(edge_direction, edge_name)
+
+        edge = self._sql_schema_info.join_descriptors[
+            self._current_classname][full_edge_name]
+
+        # we will use this to make sure that the join inside
+        # the subquery uses one alias of the current table
+        # and that the join outside the subquery uses another alias
+        join_info = (edge,
+                     outer_vertex_table,
+                     inner_vertex_type,
+                     self._current_alias)
+
+        self._fold = SQLFoldObject(inner_vertex_table,
+                                   outer_vertex_table,
+                                   inner_vertex_type,
+                                   outer_vertex_type,
+                                   join_info)
+        # find outputs for this fold in self._all_folded_outputs and add to self._fold_outputs
+        if fold_scope_location.fold_path in self._all_folded_outputs:
```

Comment:
isn't it an assertion error if you don't find this? That's what the spec of the function that computes `self._all_folded_outputs` implies.

Reply:
yep, should be changed to an assertion error instead